pkg/prom/registry.go

```diff
@@ -0,0 +1,213 @@
+package prom
+
+import (
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/gorilla/mux"
+	"github.com/peterbourgon/fastly-exporter/pkg/filter"
+	"github.com/peterbourgon/fastly-exporter/pkg/gen"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// Registry collects Prometheus metrics on a per-service basis.
+//
+// Writers (i.e. rt.Subscribers) should call MetricsFor with their specific
+// service ID, and update the returned set of Prometheus metrics. Readers (i.e.
+// Prometheus) can scrape metrics for all services via the `/metrics` endpoint.
+//
+// Prometheus's generic HTTP service discovery is supported and parameterizes on
+// service. Available services are listed on the `/sd` endpoint, and metrics for
+// a specific service can be scraped via `/metrics/{serviceID}`.
+//
+// https://prometheus.io/docs/prometheus/latest/configuration/configuration/#http_sd_config
+type Registry struct {
+	mtx              sync.Mutex
+	version          string
+	namespace        string
+	subsystem        string
+	metricNameFilter filter.Filter
+	byServiceID      map[string]*metricsRegistry
+
+	http.Handler
+}
+
+// NewRegistry returns a new and empty registry for Prometheus metrics.
+func NewRegistry(version, namespace, subsystem string, metricNameFilter filter.Filter) *Registry {
+	r := &Registry{
+		version:          version,
+		namespace:        namespace,
+		subsystem:        subsystem,
+		metricNameFilter: metricNameFilter,
+		byServiceID:      map[string]*metricsRegistry{},
+	}
+
+	router := mux.NewRouter()
+	router.StrictSlash(true)
+	router.Methods("GET").Path("/").HandlerFunc(r.handleIndex)
+	router.Methods("GET").Path("/sd").HandlerFunc(r.handleServiceDiscovery)
+	router.Methods("GET").Path("/metrics").HandlerFunc(r.handleMetrics)
+	router.Methods("GET").Path("/metrics/{serviceID:.+}").HandlerFunc(r.handleMetricsFor)
```

Comment:
To make this better follow the Prometheus conventions, we use a `__param_target` in the relabel config to send this via URL param, rather than change the `metrics_path` of the service discovery.

Something like this:
```suggestion
	router.Methods("GET").Path("/metrics").Queries("target", "{serviceID:.+}").HandlerFunc(r.handleMetricsFor)
	router.Methods("GET").Path("/metrics").HandlerFunc(r.handleMetrics)
```